### ISSUE TITLE
run unit tests in kubernetes subcomponents

### ISSUE THREAD

**Issue by liggitt** [openshift/origin#14341]
it doesn't appear that that unit tests for the staging repos (or for vendor/k8s.io/kubernetes) are running in the origin queue:

https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_check/2511/testReport/

as long as we have carry patches to these packages, we should be testing them

cc @sttts

**Comment by stevekuznetsov** [comment]
You'll want to extend the search scope for tests in [`hack/test-go.sh`](https://github.com/openshift/origin/blob/master/hack/test-go.sh#L142-L159):

```sh
if [[ -n "${test_kube}" ]]; then
    # we need to find all of the kubernetes test suites, excluding those we directly whitelisted before, the end-to-end suite, and
    # the go2idl tests which we currently do not support
    # etcd3 isn't supported yet and that test flakes upstream
    optional_kubernetes_packages="$(find vendor/k8s.io/{apimachinery,apiserver,client-go,kubernetes} -not \(                             \
      \(                                                                                          \
        -path "${kubernetes_path}/staging"                                                        \
        -o -path "${kubernetes_path}/pkg/api"                                                     \
        -o -path "${kubernetes_path}/pkg/api/v1"                                                  \
        -o -path "${kubernetes_path}/test"                                                        \
        -o -path "${kubernetes_path}/cmd/libs/go2idl/client-gen/testoutput/testgroup/unversioned" \
        -o -path "${kubernetes_path}/pkg/storage/etcd3"                                           \
        -o -path "${kubernetes_path}/third_party/golang/go/build"                                 \
      \) -prune                                                                                   \
    \) -name '*_test.go' | cut -f 2- -d / | xargs -n1 dirname | sort -u | xargs -n1 printf "./vendor/%s\n")"

    test_packages="${test_packages} ${optional_kubernetes_packages}"
fi
```

**Comment by ncdc** [comment]
Figured it out. We need to pass `-L` to `find` to get it to look at symlinks (because vendor/k8s.io/apiserver is a symlink, as are a few others). Do we want to unit test gengo, heapster, and/or metrics?

**Comment by ncdc** [comment]
Or sample-apiserver?

**Comment by sttts** [comment]
If I remember right, we only install test dependencies for k8s.io/kube. Those of gengo, heapster and metrics might be fulfilled by accident only. So better avoid those.

**Comment by liggitt** [comment]
we are apparently not running any of the kube unit tests... master has a compiler error ([missing import](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/api/v1/pod/util_test.go#L375)) in one, and CI is green

**Comment by sttts** [comment]
@liggitt we explicitly exclude `pkg/api` from the tests. Your referenced import error is in `pkg/api/v1/pod`, so those tests do not run.

Looks like these excludes even predate @stevekuznetsov's script rewrite. Does anybody remember why we introduced the excludes?

**Comment by sttts** [comment]
I double checked that the not-excluded tests run and they do, compare e.g. https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_check/3170/s3/download/scripts/test-go/logs/test-go.log. I also checked that the staging tests run and they do: search for e.g. `github.com/openshift/origin/vendor/k8s.io/kube-aggregator/pkg/controllers/status` in the upper test log.

@liggitt wdyt? We can remove the excludes when we know why we introduced them (maybe runtime or flakes?).

**Comment by liggitt** [comment]
I'd prefer we not exclude tests on things we're carrying patches on... @smarterclayton, do you know why we excluded pkg/api unit tests upstream?

**Comment by smarterclayton** [comment]
Probably a serialization round trip issue, but probably not acceptable.

**Comment by stevekuznetsov** [comment]
Added ages ago without comment: https://github.com/openshift/origin/pull/3939

**Comment by deads2k** [comment]
> Added ages ago with comment: #3939

The original pull added pkg/api to the list of tests that *always* ran.  It didn't exclude anything

**Comment by deads2k** [comment]
> The original pull added pkg/api to the list of tests that always ran. It didn't exclude anything

Oh, how much you want to bet the package we're now skipping didn't exist back then, so nothing was dropped

**Comment by liggitt** [comment]
yeah, fixed and future-proofed in #14687
